File: src/utf/thirsty_lang/ast.py

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any

from .token import Span


@dataclass
class Node:
    span: Span


@dataclass
class TypeNode(Node):
    pass


@dataclass
class NamedType(TypeNode):
    name: str


@dataclass
class GenericType(TypeNode):
    base: str
    args: list[TypeNode]


@dataclass
class FunctionType(TypeNode):
    params: list[TypeNode]
    result: TypeNode


@dataclass
class Expr(Node):
    pass


@dataclass
class LiteralExpr(Expr):
    value: Any


@dataclass
class VariableExpr(Expr):
    name: str


@dataclass
class ThisExpr(Expr):
    pass


@dataclass
class InputExpr(Expr):
    safe: bool = False


@dataclass
class ArrayExpr(Expr):
    items: list[Expr]


@dataclass
class AssignExpr(Expr):
    target: Expr
    value: Expr


@dataclass
class UnaryExpr(Expr):
    op: str
    right: Expr


@dataclass
class BinaryExpr(Expr):
    left: Expr
    op: str
    right: Expr


@dataclass
class PipeExpr(Expr):
    left: Expr
    right: Expr


@dataclass
class GuardExpr(Expr):
    condition: Expr
    when_true: Expr
    when_false: Expr | None


@dataclass
class CallExpr(Expr):
    callee: Expr
    args: list[Expr]
    safe: bool = False


@dataclass
class MemberExpr(Expr):
    obj: Expr
    name: str


@dataclass
class IndexExpr(Expr):
    obj: Expr
    index: Expr


@dataclass
class NewExpr(Expr):
    class_name: str
    args: list[Expr]


@dataclass
class AwaitExpr(Expr):
    expr: Expr


@dataclass
class CondenseExpr(Expr):
    expr: Expr


@dataclass
class EvaporateExpr(Expr):
    expr: Expr


@dataclass
class Stmt(Node):
    pass


@dataclass
class BlockStmt(Stmt):
    statements: list[Stmt]


@dataclass
class ExprStmt(Stmt):
    expr: Expr


@dataclass
class PrintStmt(Stmt):
    expr: Expr
    safe: bool = False


@dataclass
class ReturnStmt(Stmt):
    expr: Expr | None


@dataclass
class ThrowStmt(Stmt):
    expr: Expr


@dataclass
class DripStmt(Stmt):
    name: str
    amount: Expr | None = None


@dataclass
class IfStmt(Stmt):
    condition: Expr
    then_branch: BlockStmt
    else_branch: BlockStmt | None


@dataclass
class LoopStmt(Stmt):
    count: Expr
    body: BlockStmt


@dataclass
class CatchClause(Node):
    name: str
    type_name: str
    block: BlockStmt


@dataclass
class TryStmt(Stmt):
    try_block: BlockStmt
    catches: list[CatchClause]
    finally_block: BlockStmt | None = None


@dataclass
class Param(Node):
    name: str
    type_node: TypeNode


@dataclass
class VarDecl(Stmt):
    name: str
    type_node: TypeNode
    initializer: Expr
    mutable: bool = False
    visibility: str | None = None
    is_field: bool = False


@dataclass
class FunctionDecl(Stmt):
    name: str
    params: list[Param]
    return_type: TypeNode | None
    body: BlockStmt
    is_async: bool = False
    visibility: str | None = None
    is_method: bool = False


@dataclass
class ClassDecl(Stmt):
    name: str
    members: list[Stmt]


@dataclass
class ImportDecl(Stmt):
    module: str
    alias: str | None = None


@dataclass
class Program(Node):
    declarations: list[Stmt] = field(default_factory=list)

File: src/thirsty_lang/src/ast.js
'use strict';

/**
 * Thirsty-Lang AST Node Classes
 * Matching Python ast.py
 */

class Node {
  constructor(span) {
    this.span = span;
  }
}

// Type nodes
class TypeNode extends Node {}

class NamedType extends TypeNode {
  constructor(span, name) {
    super(span);
    this.name = name;
  }
}

class GenericType extends TypeNode {
  constructor(span, base, args) {
    super(span);
    this.base = base;
    this.args = args;
  }
}

class FunctionType extends TypeNode {
  constructor(span, params, result) {
    super(span);
    this.params = params;
    this.result = result;
  }
}

// Expressions
class Expr extends Node {}

class LiteralExpr extends Expr {
  constructor(span, value) {
    super(span);
    this.value = value;
  }
}

class VariableExpr extends Expr {
  constructor(span, name) {
    super(span);
    this.name = name;
  }
}

class ThisExpr extends Expr {
  constructor(span) {
    super(span);
  }
}

class InputExpr extends Expr {
  constructor(span, safe = false) {
    super(span);
    this.safe = safe;
  }
}

class ArrayExpr extends Expr {
  constructor(span, items) {
    super(span);
    this.items = items;
  }
}

class AssignExpr extends Expr {
  constructor(span, target, value) {
    super(span);
    this.target = target;
    this.value = value;
  }
}

class UnaryExpr extends Expr {
  constructor(span, op, right) {
    super(span);
    this.op = op;
    this.right = right;
  }
}

class BinaryExpr extends Expr {
  constructor(span, left, op, right) {
    super(span);
    this.left = left;
    this.op = op;
    this.right = right;
  }
}

class PipeExpr extends Expr {
  constructor(span, left, right) {
    super(span);
    this.left = left;
    this.right = right;
  }
}

class GuardExpr extends Expr {
  constructor(span, condition, whenTrue, whenFalse) {
    super(span);
    this.condition = condition;
    this.whenTrue = whenTrue;
    this.whenFalse = whenFalse;
  }
}

class CallExpr extends Expr {
  constructor(span, callee, args, safe = false) {
    super(span);
    this.callee = callee;
    this.args = args;
    this.safe = safe;
  }
}

class MemberExpr extends Expr {
  constructor(span, obj, name) {
    super(span);
    this.obj = obj;
    this.name = name;
  }
}

class IndexExpr extends Expr {
  constructor(span, obj, index) {
    super(span);
    this.obj = obj;
    this.index = index;
  }
}

class NewExpr extends Expr {
  constructor(span, className, args) {
    super(span);
    this.className = className;
    this.args = args;
  }
}

class AwaitExpr extends Expr {
  constructor(span, expr) {
    super(span);
    this.expr = expr;
  }
}

class CondenseExpr extends Expr {
  constructor(span, expr) {
    super(span);
    this.expr = expr;
  }
}

class EvaporateExpr extends Expr {
  constructor(span, expr) {
    super(span);
    this.expr = expr;
  }
}

// Statements
class Stmt extends Node {}

class BlockStmt extends Stmt {
  constructor(span, statements) {
    super(span);
    this.statements = statements;
  }
}

class ExprStmt extends Stmt {
  constructor(span, expr) {
    super(span);
    this.expr = expr;
  }
}

class PrintStmt extends Stmt {
  constructor(span, expr, safe = false) {
    super(span);
    this.expr = expr;
    this.safe = safe;
  }
}

class ReturnStmt extends Stmt {
  constructor(span, expr) {
    super(span);
    this.expr = expr; // null if no value
  }
}

class ThrowStmt extends Stmt {
  constructor(span, expr) {
    super(span);
    this.expr = expr;
  }
}

class DripStmt extends Stmt {
  constructor(span, name, amount = null) {
    super(span);
    this.name = name;
    this.amount = amount;
  }
}

class IfStmt extends Stmt {
  constructor(span, condition, thenBranch, elseBranch) {
    super(span);
    this.condition = condition;
    this.thenBranch = thenBranch;
    this.elseBranch = elseBranch;
  }
}

class LoopStmt extends Stmt {
  constructor(span, count, body) {
    super(span);
    this.count = count;
    this.body = body;
  }
}

class CatchClause extends Node {
  constructor(span, name, typeName, block) {
    super(span);
    this.name = name;
    this.typeName = typeName;
    this.block = block;
  }
}

class TryStmt extends Stmt {
  constructor(span, tryBlock, catches, finallyBlock = null) {
    super(span);
    this.tryBlock = tryBlock;
    this.catches = catches;
    this.finallyBlock = finallyBlock;
  }
}

class Param extends Node {
  constructor(span, name, typeNode) {
    super(span);
    this.name = name;
    this.typeNode = typeNode;
  }
}

class VarDecl extends Stmt {
  constructor(span, name, typeNode, initializer, mutable = false, visibility = null, isField = false) {
    super(span);
    this.name = name;
    this.typeNode = typeNode;
    this.initializer = initializer;
    this.mutable = mutable;
    this.visibility = visibility;
    this.isField = isField;
  }
}

class FunctionDecl extends Stmt {
  constructor(span, name, params, returnType, body, isAsync = false, visibility = null, isMethod = false) {
    super(span);
    this.name = name;
    this.params = params;
    this.returnType = returnType;
    this.body = body;
    this.isAsync = isAsync;
    this.visibility = visibility;
    this.isMethod = isMethod;
  }
}

class ClassDecl extends Stmt {
  constructor(span, name, members) {
    super(span);
    this.name = name;
    this.members = members;
  }
}

class ImportDecl extends Stmt {
  constructor(span, module, alias = null) {
    super(span);
    this.module = module;
    this.alias = alias;
  }
}

class ModuleHeader extends Node {
  constructor(span, name, mode = 'core') {
    super(span);
    this.name = name;
    this.mode = mode;
  }
}

class RequiresClause extends Node {
  constructor(span, annotation) {
    super(span);
    this.annotation = annotation;
  }
}

class GovernedFunctionDecl extends Stmt {
  constructor(span, name, params, returnType, body, requires = [], isAsync = false, visibility = null, isMethod = false) {
    super(span);
    this.name = name;
    this.params = params;
    this.returnType = returnType;
    this.body = body;
    this.requires = requires;
    this.isAsync = isAsync;
    this.visibility = visibility;
    this.isMethod = isMethod;
  }
}

class EnumDecl extends Stmt {
  constructor(span, name, variants) {
    super(span);
    this.name = name;
    this.variants = variants;
  }
}

class StructDecl extends Stmt {
  constructor(span, name, fields) {
    super(span);
    this.name = name;
    this.fields = fields;
  }
}

class InterfaceDecl extends Stmt {
  constructor(span, name, methods) {
    super(span);
    this.name = name;
    this.methods = methods;
  }
}

class Program extends Node {
  constructor(span, declarations = [], header = null) {
    super(span);
    this.declarations = declarations;
    this.header = header;
  }
}

class MutationDecl extends Stmt {
  constructor(span, name, params, shadowBlock, invariantBlock, canonicalBlock) {
    super(span);
    this.name = name;
    this.params = params;
    this.shadowBlock = shadowBlock;
    this.invariantBlock = invariantBlock;
    this.canonicalBlock = canonicalBlock;
  }
}

module.exports = {
  Node, TypeNode, NamedType, GenericType, FunctionType,
  Expr, LiteralExpr, VariableExpr, ThisExpr, InputExpr, ArrayExpr,
  AssignExpr, UnaryExpr, BinaryExpr, PipeExpr, GuardExpr,
  CallExpr, MemberExpr, IndexExpr, NewExpr, AwaitExpr,
  CondenseExpr, EvaporateExpr,
  Stmt, BlockStmt, ExprStmt, PrintStmt, ReturnStmt, ThrowStmt,
  DripStmt, IfStmt, LoopStmt, CatchClause, TryStmt,
  Param, VarDecl, FunctionDecl, ClassDecl, ImportDecl,
  ModuleHeader, RequiresClause, GovernedFunctionDecl,
  EnumDecl, StructDecl, InterfaceDecl, Program, MutationDecl,
};
